matrix/client/query_llm.py

```diff
@@ -287,17 +311,26 @@ async def make_request(
                                 "response_timestamp": time.time(),
                             },
                         }
-                        if logprobs and response.choices[0].logprobs is not None:  # type: ignore[attr-defined]
-                            lp = [
-                                [
-                                    {"token": elem[0], "logprob": elem[1]}
-                                    for elem in zip(
-                                        response.choices[i].logprobs.tokens,  # type: ignore[union-attr]
-                                        response.choices[i].logprobs.token_logprobs,  # type: ignore[union-attr]
-                                    )  # type: ignore[attr-defined]
-                                ]
-                                for i in range(n)
-                            ]
+                        if (logprobs or top_logprobs is not None) and response.choices[
+                            0
+                        ].logprobs is not None:  # type: ignore[attr-defined]
+                            lp = []
+                            for i in range(n):
+                                tokens = response.choices[i].logprobs.tokens  # type: ignore[union-attr]
+                                token_lps = response.choices[i].logprobs.token_logprobs  # type: ignore[union-attr]
+                                top_lps_list = response.choices[i].logprobs.top_logprobs  # type: ignore[union-attr]
+                                current = []
+                                for idx, (token, lp_token) in enumerate(
+                                    zip(tokens, token_lps)
+                                ):
+                                    entry = {"token": token, "logprob": lp_token}
+                                    if top_lps_list:
+                                        entry["top_logprobs"] = [
+                                            {"token": t, "logprob": tl}
+                                            for t, tl in top_lps_list[idx].items()
+                                        ]
+                                    current.append(entry)
+                                lp.append(current)
                             result["response"]["logprobs"] = lp
```

Comment:
wondering if it is possible to follow the same coding structure as in line 459-477?